### PR TITLE
Update to spec v0.8.3

### PR DESCRIPTION
**Pull request by 0xKiwi**
Part of #3144

This PR aims to updates the Prysm client to spec v0.8.3, it includes the following:

* fix start shard for compact committees root 
* ignore latest messages in fork choice prior to latest justified 
* 0.8.3 spec tests

### CODE REVIEW

File: beacon-chain/core/helpers/committee.go [hunk 1]
```diff
@@ -175,6 +175,17 @@ func AttestingIndices(state *pb.BeaconState, data *ethpb.AttestationData, bf bit
 	return indices, nil
 }
 
+// VerifyBitfield validates a bitfield with a given committee size.
+func VerifyBitfield(bf bitfield.Bitfield, committeeSize uint64) error {
```
Comment (nisdas) [inline beacon-chain/core/helpers/committee.go]: this isnt verifying the bitfield, its just checking the length of the bitfield. Maybe change the naming to reflect this more

File: proto/testing/ssz_static_test.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+					)
+				}
+			})
+		}
+	}
+}
+
+func UnmarshalledSSZ(serializedBytes []byte, folderName string) (interface{}, error) {
+	var obj interface{}
+	if folderName == "Attestation" {
```
Comment (nisdas) [inline proto/testing/ssz_static_test.go]: there are better ways than creating a 50 line if else block. You can use a switch and case for this

Comment (nisdas) [reply proto/testing/ssz_static_test.go]: Also add in a case for the default when none of it matches

File: beacon-chain/core/blocks/block_operations.go [hunk 3]
```diff
@@ -564,6 +570,24 @@ func ProcessAttestation(beaconState *pb.BeaconState, att *ethpb.Attestation) (*p
...
+
+	if data.Crosslink.Shard > params.BeaconConfig().ShardCount {
+		return nil, fmt.Errorf(
+			"expected crosslink shard %d to be less than SHARD_COUNT %d",
+			data.Crosslink.Shard,
+			params.BeaconConfig().ShardCount,
+		)
+	}
+
+	if !(data.Target.Epoch == helpers.PrevEpoch(beaconState) || data.Target.Epoch == helpers.CurrentEpoch(beaconState)) {
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block_operations.go]:
```suggestion
	if data.Target.Epoch != helpers.PrevEpoch(beaconState) && data.Target.Epoch != helpers.CurrentEpoch(beaconState) {
```

File: beacon-chain/core/blocks/block_operations.go [hunk 4]
```diff
@@ -564,6 +570,24 @@ func ProcessAttestation(beaconState *pb.BeaconState, att *ethpb.Attestation) (*p
...
+		return nil, fmt.Errorf(
+			"expected target epoch %d == %d or %d",
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block_operations.go]:
```suggestion
			"expected target epoch (%d) to be in the previous epoch (%d) or current epoch (%d)",
```

File: beacon-chain/core/blocks/block_operations.go [hunk 5]
```diff
@@ -586,6 +610,11 @@ func ProcessAttestationNoVerify(beaconState *pb.BeaconState, att *ethpb.Attestat
 			params.BeaconConfig().SlotsPerEpoch,
 		)
 	}
+
+	if err := helpers.VerifyAttestationBitfields(beaconState, att); err != nil {
+		return nil, fmt.Errorf("could not verify attestation bitfields: %v", err)
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/block_operations.go]:
```suggestion
		return nil, errors.Wrap(err, "could not verify attestation bitfields")
```

File: beacon-chain/core/helpers/committee.go [hunk 6]
```diff
@@ -346,6 +357,27 @@ func StartShard(state *pb.BeaconState, epoch uint64) (uint64, error) {
...
+	if err != nil {
+		return errors.Wrap(err, "could not retrieve crosslink committees")
+	}
+
+	if committee == nil {
+		return fmt.Errorf("no committee exist for shard in the attestation")
+	}
+
+	if err := VerifyBitfield(att.AggregationBits, uint64(len(committee))); err != nil {
+		return fmt.Errorf("failed to verify aggregation bitfield: %v", err)
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/committee.go]:
```suggestion
		return errors.Wrap(err, "failed to verify aggregation bitfield")
```

File: beacon-chain/core/helpers/committee.go [hunk 7]
```diff
@@ -346,6 +357,27 @@ func StartShard(state *pb.BeaconState, epoch uint64) (uint64, error) {
...
+// VerifyAttestationBitfields verifies that an attestations aggregation and custody bitfield is valid in respect
+// to the committee at that slot.
+func VerifyAttestationBitfields(bState *pb.BeaconState, att *ethpb.Attestation) error {
+	committee, err := CrosslinkCommittee(bState, att.Data.Target.Epoch, att.Data.Crosslink.Shard)
+	if err != nil {
+		return errors.Wrap(err, "could not retrieve crosslink committees")
+	}
+
+	if committee == nil {
+		return fmt.Errorf("no committee exist for shard in the attestation")
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/committee.go]:
```suggestion
		return errors.New("no committee exist for shard in the attestation")
```

File: beacon-chain/core/helpers/committee.go [hunk 8]
```diff
@@ -346,6 +357,27 @@ func StartShard(state *pb.BeaconState, epoch uint64) (uint64, error) {
...
+	}
+
+	if err := VerifyBitfield(att.AggregationBits, uint64(len(committee))); err != nil {
+		return fmt.Errorf("failed to verify aggregation bitfield: %v", err)
+	}
+	if err := VerifyBitfield(att.CustodyBits, uint64(len(committee))); err != nil {
+		return fmt.Errorf("failed to verify custody bitfield: %v", err)
```
Comment (prestonvanloon) [inline beacon-chain/core/helpers/committee.go]:
```suggestion
		return errors.Wrap(err, "failed to verify custody bitfield")
```

File: beacon-chain/core/helpers/committee.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/core/helpers/committee.go]: `VerifyBitfieldLength`

File: proto/testing/ssz_static_test.go (hunk 2 above)
Comment (prestonvanloon) [reply proto/testing/ssz_static_test.go]: +1 switch statement

**Comment by prestonvanloon** [comment]
Working at runtime with https://github.com/prysmaticlabs/eth1-mock-rpc/pull/27

**Comment by prestonvanloon** [comment]
There is one test that fails with minimal config, but we'll follow up in another PR

**Pull request merged**